operator/src/main/java/org/bf2/operator/managers/MetricsManager.java

```diff
@@ -0,0 +1,121 @@
+    static final String KAFKA_INSTANCE_CONNECTION_LIMIT = "kafka_instance_connection_limit";
+    static final String KAFKA_INSTANCE_CONNECTION_CREATION_RATE_LIMIT = "kafka_instance_connection_creation_rate_limit";
+
+    static final String TAG_LABEL_OWNER = "owner";
+    static final String TAG_LABEL_BROKER_ID = "broker_id";
+    static final String TAG_LABEL_NAMESPACE = "namespace";
+    static final String TAG_LABEL_INSTANCE_NAME = "instance_name";
+
+    static final Tag OWNER = Tag.of(TAG_LABEL_OWNER, "KafkaInstanceMetricsManager");
+    static final String TAG_LABEL_LISTENER = "listener";
+
+    @Inject
+    ResourceInformerFactory resourceInformerFactory;
+
+    @Inject
+    MeterRegistry meterRegistry;
+
+    @Inject
+    KubernetesClient kubernetesClient;
+
+    private volatile ResourceInformer<Kafka> kafkaResourceInformer;
+
+    protected synchronized void createInformer() {
+        if (kafkaResourceInformer != null) {
+            return;
+        }
+        kafkaResourceInformer = resourceInformerFactory.create(Kafka.class, filter(kubernetesClient.resources(Kafka.class, KafkaList.class)), new ResourceEventHandler<>() {
+            @Override
+            public void onAdd(Kafka kafka) {
+                createMetrics(kafka);
+            }
+
+            @Override
+            public void onUpdate(Kafka oldObj, Kafka newObj) {
+                createMetrics(newObj);
+            }
+
+            @Override
+            public void onDelete(Kafka obj, boolean deletedFinalStateUnknown) {
+                Search.in(meterRegistry).tags(buildKafkaInstanceTags(obj)).meters().forEach(meterRegistry::remove);
+
+            }
+
+            private void createMetrics(Kafka kafka) {
+                Tags tags = buildKafkaInstanceTags(kafka);
+                meterRegistry.gauge(KAFKA_INSTANCE_BROKERS_DESIRED_COUNT, tags, kafka, this::replicas);
+                meterRegistry.gauge(KAFKA_INSTANCE_PARTITION_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MAX_PARTITIONS));
+                meterRegistry.gauge(KAFKA_INSTANCE_MAX_MESSAGE_SIZE_LIMIT, tags, kafka, k -> kafkaConfigValue(k, KafkaCluster.MESSAGE_MAX_BYTES));
+
+                if (kafka.getSpec() != null && kafka.getSpec().getKafka() != null) {
+                    IntStream.range(0, kafka.getSpec().getKafka().getReplicas()).forEach(ordinal -> {
+                        Tags brokerTags = Tags.concat(tags, Tags.of(Tag.of(TAG_LABEL_BROKER_ID, String.valueOf(ordinal))));
+
+                        if (kafka.getSpec().getKafka().getListeners() != null) {
+                            kafka.getSpec().getKafka().getListeners().stream().filter(l -> EXTERNAL_LISTENER_NAME.equals(l.getName())).forEach(l -> {
+                                String listenerName = String.format("%s-%d", l.getName().toUpperCase(Locale.ROOT), l.getPort());
+                                Tags listenerTags = Tags.concat(brokerTags, Tags.of(Tag.of(TAG_LABEL_LISTENER, listenerName)));
```

Comment:
Maybe a small comment here to give the reason for the `toUpperCase` since it's outside of fleetshard.

Reply:
Done